router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+		return fwd.handleControlMsg(special.frame)
+	}
+
+	// The special frame types are distinguished by length
+	switch len(special.frame) {
+	case EthernetOverhead + 8:
+		return fwd.handleHeartbeat(special)
+
+	case FragTestSize:
+		return fwd.handleFragTest(special.frame)
+
+	default:
+		return fwd.handlePMTUTest(special.frame)
+	}
+}
+
+const (
+	HeartbeatAck = iota
+	FragTestAck
+	PMTUTestAck
+)
+
+func (fwd *sleeveForwarder) handleControlMsg(msg []byte) error {
+	if len(msg) == 0 {
+		log.Println(fwd.logPrefix(),
+			"Received zero-length control message")
+		return nil
+	}
+
+	switch msg[0] {
+	case HeartbeatAck:
+		return fwd.handleHeartbeatAck()
+
+	case FragTestAck:
+		return fwd.handleFragTestAck()
+
+	case PMTUTestAck:
+		return fwd.handlePMTUTestAck(msg)
+
+	default:
+		log.Println(fwd.logPrefix(),
+			"Ignoring unknown control message:", msg[0])
+		return nil
+	}
+}
+
+func (fwd *sleeveForwarder) confirmed() error {
+	log.Debug(fwd.logPrefix(), "confirmed")
+
+	if fwd.heartbeatInterval != 0 {
+		// already confirmed
+		return nil
+	}
+
+	// heartbeatInterval flags that we want to send heartbeats,
+	// even if we don't do sendHeartbeat() yet due to lacking the
+	// remote address.
+	fwd.heartbeatInterval = time.Duration(FastHeartbeat)
+	if fwd.remoteAddr != nil {
+		if err := fwd.sendHeartbeat(); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout = time.NewTimer(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) sendHeartbeat() error {
+	log.Debug(fwd.logPrefix(), "sendHeartbeat")
+
+	// Prime the timer for the next heartbeat.  We don't use a
+	// ticker because the interval is not constant.
+	fwd.heartbeatTimer = setTimer(fwd.heartbeatTimer, fwd.heartbeatInterval)
+
+	buf := make([]byte, EthernetOverhead+8)
+	binary.BigEndian.PutUint64(buf[EthernetOverhead:], fwd.connUID)
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF, buf)
+}
+
+func (fwd *sleeveForwarder) handleHeartbeat(special specialFrame) error {
+	uid := binary.BigEndian.Uint64(special.frame[EthernetOverhead:])
+	if uid != fwd.connUID {
+		return nil
+	}
+
+	log.Debug(fwd.logPrefix(), "handleHeartbeat")
+
+	if fwd.remoteAddr == nil {
+		fwd.setRemoteAddr(special.sender)
+		if fwd.heartbeatInterval != time.Duration(0) {
+			if err := fwd.sendHeartbeat(); err != nil {
+				return err
+			}
+		}
+	} else if !udpAddrsEqual(fwd.remoteAddr, special.sender) {
+		log.Println(fwd.logPrefix(),
+			"Peer UDP address changed to", special.sender)
+		fwd.setRemoteAddr(special.sender)
+	}
+
+	if !fwd.ackedHeartbeat {
+		fwd.ackedHeartbeat = true
+		if err := fwd.sendControlMsg([]byte{HeartbeatAck}); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout.Reset(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) setRemoteAddr(addr *net.UDPAddr) {
+	// Although we don't need to lock when reading remoteAddr,
+	// because this thread is the only one that modifies
+	// remoteAddr, we do need to lock when writing it, because
+	// memory models.
```

Comment:
I'm mystified by this comment. Firstly, it's not clear what 'this thread' is referring to; secondly, 'because memory models', whilst amusing, is insufficiently explicatory :smile: